hazelcast/src/hazelcast/client/network.cpp

```diff
@@ -183,416 +208,272 @@ namespace hazelcast {
 
             std::shared_ptr<Connection>
             ClientConnectionManagerImpl::getOrConnect(const Address &address) {
-                return getOrConnect(address, false);
-            }
+                auto connection = getConnection(address);
+                if (connection) {
+                    return connection;
+                }
 
-            std::shared_ptr<Connection> ClientConnectionManagerImpl::getOwnerConnection() {
-                std::shared_ptr<Address> address = ownerConnectionAddress;
-                if (address.get() == NULL) {
-                    return std::shared_ptr<Connection>();
+                auto future = std::make_shared<ConnectionFuture>(address, connection, connectionsInProgress);
+                auto f = connectionsInProgress.putIfAbsent(address, future);
+                bool outstanding_connection_exists = static_cast<bool>(f);
```

Comment:
Are we casting a `shared_ptr<ConnectionFuture>` to boolean?  Should we get if f.get() is null or not ?

Reply:
casting to bool is the same as `f.get() == nullptr`.

Reply:
It seems you have used the same trick also here.
```
connection = getConnection(address);
                if (connection) {
```
That means shared ptr overrides bool operator

Reply:
yes, exactly shared_ptr and boost::optional both has this bool operator, and it is less code.